academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1623 @@
+
+            # Open a new FTP connection
+            ftp_conn = FTP()
+            ftp_conn.connect(host=ftp_server_url, port=ftp_port)
+            ftp_conn.login()  # anonymous login (publicly available data)
+
+            logging.info(
+                f"FTP connection to Pubmed's servers has been reset after {reset_ftp_counter} downloads to avoid issues."
+            )
+
+        download_attempt_count = 1
+        download_success = False
+        while download_attempt_count <= max_download_retry and not download_success:
+            logging.info(f"Downloading: {datafile.filename} Attempt: {download_attempt_count}")
+            try:
+                # Download file
+                with open(datafile.download_file_path, "wb") as f:
+                    ftp_conn.retrbinary(f"RETR {datafile.path_on_ftp}", f.write)
+                logging.info(f"File downloaded to - {datafile.download_file_path}")
+            except error_reply:
+                logging.info(f"Unable to download {datafile.path_on_ftp} from PubMed's FTP server {ftp_server_url}.")
+
+            # Create the hash from the above downloaded file.
+            with open(datafile.download_file_path, "rb") as f_hash:
+                data = f_hash.read()
+                md5hash_from_download = hashlib.md5(data).hexdigest()
+
+            logging.info(f"MD5 hash: {md5hash_from_download} for downloaded file: {datafile.filename}")
+
+            # Need to have a download catch for the hash file as well, otherwise it can break the download loop.
+            try:
+                # Download corresponding md5 hash.
+                with open(f"{datafile.download_file_path}.md5", "wb") as f:
+                    ftp_conn.retrbinary(f"RETR {datafile.path_on_ftp}.md5", f.write)
+            except error_reply:
+                logging.info(
+                    f"Unable to download {datafile.path_on_ftp}.md5 from PubMed's FTP server {ftp_server_url}."
+                )
+
+            with open(f"{datafile.download_file_path}.md5", "r") as f_md5:
+                md5_from_pubmed_ftp = f_md5.read()
+
+            # If md5 does not match, retry download.
+            if md5hash_from_download in md5_from_pubmed_ftp:
+                download_success = True
+            else:
+                logging.info(
+                    f"MD5 hash does not match the given checksum from server: {datafile.download_file_path}\
+                            - Retrying download ..."
+                )
+
+            download_attempt_count += 1
+
+        assert (
+            download_success
+        ), f"Unable to download {datafile.download_file_path} from PubMed's FTP server \
+                    {ftp_server_url} after {max_download_retry} tries."
+
+    # Close the FTP connection after downloading the required files.
+    ftp_conn.close()
+
+    return download_success
+
+
+def transform_pubmed_xml_file_to_jsonl(datafile: Datafile, entity_list: List[PubmedEntity]) -> Union[Datafile, bool]:
```

Comment:
I would give it a shorter name, e.g. `transform_pubmed` and split this function out into a few different functions, e.g. as below. I think it is easier to understand what is going on just by directly setting the dictionary keys / making explicit functions to get the articles / deletes out of the data.

```python
def load_datafile(file_path: str):
    logging.info(f"Reading in file - {file_path}")

    with gzip.open(file_path, "rb") as f_in:
        # Use the BioPython package for reading in the Pubmed XML files.
        # This package also checks against its own DTD schema defined in the XML header.
        data = Entrez.read(f_in, validate=True)

    # Need pull out XML attributes from the Biopython data classes.
    data = add_attributes_to_data_from_biopython_classes(data)

    #  Remove unwanted nested list structure from the Pubmed dictionary.
    data = change_pubmed_list_structure(data)

    return data


def save_pubmed_jsonl(file_path: str, data: List[dict]):
    if file_path.endswith(".gz"):
        with gzip.open(file_path, "w") as f_out:
            for line in data:
                f_out.write(str.encode(json.dumps(line, cls=PubMedCustomEncoder) + "\n"))
    else:
        with open(file_path, "w") as f_out:
            for line in data:
                f_out.write(json.dumps(line, cls=PubMedCustomEncoder) + "\n")


def parse_articles(data: Dict):
    try:
        return data["PubmedArticle"]
    except KeyError:
        logging.info(f"No PubmedArticle records in file")
        return []


def parse_deletes(data: Dict):
    try:
         return data["DeleteCitation"]["PMID"]
    except KeyError:
        logging.info(f"No DeleteCitation.PMID records in file")
        return []


def transform_pubmed(input_path: str, upsert_path: str, delete_path: Optional[str] = None) -> Union[Datafile, bool]:
    """
    Convert a single Pubmed XML file to JSONL, pulling out any of the Pubmed entities and their upserts and or deletes.
    Used in parallelised transform section.

    :return: Datafile that's been transformed.
    """

    logging.info(f"Reading in file - {input_path}")

    try:
        data = load_datafile(input_path)
    except ValidationError:
        logging.info(f"Fields in XML are not valid against it's own DTD file - {input_path}")
        return False

    # Save upserts
    upserts = parse_articles(data)
    save_pubmed_jsonl(upsert_path, upserts)

    # Save deletes
    if delete_path is not None:
        deletes = parse_deletes(data)
        save_jsonl(delete_path, deletes)

    return input_path
```